### PR TITLE
[v1.33.0] Update package manager to add newly found packages to the `packages.json`

### LINKED ISSUE #609: aea packages lock should add entry for non-existent packages in packages.json
**Is your feature request related to a problem? Please describe.**

If there is a package defined within the structure of the local repository, but NOT an entry in `packages.json`, upon callin `aea packages lock` it will fail indicating that the entry is not present. See for example conversation in:

https://github.com/valory-xyz/open-autonomy/pull/1863#discussion_r1148795073

**Describe the solution you'd like**
Whenever `aea packages lock` finds a folder within the local registry, it should prompt the user if he wishes to add this entry to packages.json.

It should display the information (package folder, author name, etc...) when it asks the user to include it in packages.json.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by angrybayblade**
## Proposed changes

- Updates package manager to add newly found packages to the `packages.json` instead of raising error

## Fixes

#609 

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes and CI passes too
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by angrybayblade** [comment]
@jmoreira-valory With this update the `autonomy packages lock` will add new packages to the `packages.json` file instead of raising errors.

If the author name of the package is same as the author name from the cli config (`~/.aea/cli_config.yaml`) it'll be added as a dev package and if not then it'll be added as a third party package

File: aea/cli/packages.py [hunk 1]
```diff
@@ -48,6 +48,19 @@ class SyncTypes:  # pylint: disable=too-few-public-methods
     ALL = "all"
 
 
+def package_type_selector_prompt() -> str:
```
Comment (solarw) [inline aea/cli/packages.py:51]: probably provide extra args like pacakge name or path to show in a prompt?

File: tests/test_package_manager/test_v1.py [hunk 2]
```diff
@@ -408,17 +406,36 @@ def test_update_package_hashes(
...
+        ), mock.patch.object(
+            pm,
+            "update_dependencies",
         ):
-            with pytest.raises(
-                PackageNotValid,
-                match="Found a package which is not listed in the `packages.json`",
-            ):
-                pm.update_package_hashes()
+            pm.update_package_hashes(selector_prompt=lambda: "dev")
```
Comment (solarw) [inline tests/test_package_manager/test_v1.py:435]: probably use mock as selector_prompt, and check it was actually called once?

File: tests/test_package_manager/test_v1.py [hunk 3]
```diff
@@ -438,7 +455,7 @@ def test_update_package_hashes(self, caplog) -> None:
         pm = PackageManagerV1.from_dir(self.packages_dir_path)
 
         with caplog.at_level(logging.WARNING):
-            pm.update_package_hashes().dump()
+            pm.update_package_hashes(selector_prompt=lambda: "third_party").dump()
```
Comment (solarw) [inline tests/test_package_manager/test_v1.py:458]: probably use mock as selector_prompt, and check it was actually called once?

**Comment by solarw** [comment]
not sure i got the final use case.
is it going to be added to cli command to help user to add new packages?
if i want run automation unattended?

**Review by solarw — APPROVED** [review]
LGTM, minor comments

**Comment by angrybayblade** [comment]
> not sure i got the final use case.

The `aea packages lock` used to raise error if a new package was found which is not defined in `packages.json`, with this update the new package will be added to `packages.json`

> is it going to be added to cli command to help user to add new packages?

Yes

> if i want run automation unattended?

I don't understand, do you mean running lock command in CI?

**Comment by solarw** [comment]
> I don't understand, do you mean running lock command in CI?

if I'm going to use packages code in some script and dont want to interact with user.

Suppose it's better to have some options like:
* raise error - to have the same behaviour as before
* skip package - it's not in registry and new - do nothing
* add to dev
* add to 3rd

user selectable are: skip, add to dev, add to 3rd
raise error can be used for own code, if we need to be sure no extra packages are present.

next point: package details argument  to prompt function: self contained solution does not rely on caller's code

**Comment by jmoreira-valory** [comment]
> next point: package details argument to prompt function: self contained solution does not rely on caller's code

I don't know if I got you right, @solarw :

You suggest that
1. the default behaviour when a user uses the CLI `packages lock` is to show the prompt when an unexpected package is found ("add 3party, add dev, ignore")
2. add some sort of flag (e.g., --halt-on-error)  to be used on scripts without user interaction.

is this what you meant?

Also, as pointed by @angrybayblade "ignoring" a package will cause the issue that the local registry is in a "bad state", and it will be prompted on each call to the CLI.

File: aea/package_manager/v1.py [hunk 4]
```diff
@@ -272,7 +273,9 @@ def sync(
         return self
 
     def update_package_hashes(
-        self, selector_prompt: Callable[[], str]
+        self,
+        selector_prompt: Optional[Callable[[], str]] = None,
+        skip_missing: bool = False,
```
Comment (angrybayblade) [inline aea/package_manager/v1.py:278]: @solarw Selector prompt is optional, if you don't provide a selector prompt callable it'll raise error if you don't set `skip_missing` to `True` explicitly.

Comment (angrybayblade) [reply aea/package_manager/v1.py:278]: On the cli, you can provide `--skip-missing` flag to skip the packages which are missing from the `packages.json`

Comment (solarw) [reply aea/package_manager/v1.py:278]: @angrybayblade nice, thanks!

**Review by solarw — APPROVED** [review]
LGTM

**Review by jmoreira-valory — APPROVED** [review]

**Pull request merged**